File: python/Heap or Priority Queue/find_median_data_stream.py
import heapq

class MedianFinder:
    def __init__(self):
        # Max-heap for the smaller half (invert values for heapq)
        self.small = []
        # Min-heap for the larger half
        self.large = []

    def addNum(self, num: int) -> None:
        """
        Adds a number into the data structure.
        Approach:
        - Use two heaps:
          - self.small: Max-heap (as min-heap with negative values) for the smaller half
          - self.large: Min-heap for the larger half
        - Always push to self.small first, then balance heaps so:
          - All elements in self.small <= all elements in self.large
          - The size difference between heaps is at most 1
        Time Complexity: O(log n) per insertion (heap push/pop)
        Space Complexity: O(n) for storing all numbers
        """
        # Push to max-heap (invert value for heapq)
        heapq.heappush(self.small, -num)

        # Ensure every element in self.small <= every element in self.large
        if self.small and self.large and (-self.small[0] > self.large[0]):
            val = -heapq.heappop(self.small)
            heapq.heappush(self.large, val)
        
        # Balance the sizes (small can have at most 1 more than large)
        if len(self.small) > len(self.large) + 1:
            val = -heapq.heappop(self.small)
            heapq.heappush(self.large, val)
        if len(self.large) > len(self.small) + 1:
            val = heapq.heappop(self.large)
            heapq.heappush(self.small, -val)

    def findMedian(self) -> float:
        """
        Returns the median of all elements so far.
        Time Complexity: O(1)
        Space Complexity: O(1) (not counting the data stored)
        """
        if len(self.small) > len(self.large):
            return -self.small[0]
        if len(self.large) > len(self.small):
            return self.large[0]
        # If even number of elements, median is the average of the two middle values
        return ((-self.small[0]) + self.large[0]) / 2

# -------------------
# Test Cases
# -------------------
def test_median_finder():
    mf = MedianFinder()
    # Test case 1: Add numbers and check median
    mf.addNum(1)
    assert mf.findMedian() == 1
    mf.addNum(2)
    assert mf.findMedian() == 1.5
    mf.addNum(3)
    assert mf.findMedian() == 2
    # Test case 2: Add more numbers
    mf2 = MedianFinder()
    nums2 = [5, 15, 1, 3]
    expected2 = [5, 10.0, 5, 4.0]
    for i, num in enumerate(nums2):
        mf2.addNum(num)
        assert mf2.findMedian() == expected2[i]
    # Test case 3: All same number
    mf3 = MedianFinder()
    for _ in range(5):
        mf3.addNum(7)
    assert mf3.findMedian() == 7
    # Test case 4: Negative numbers
    mf4 = MedianFinder()
    mf4.addNum(-1)
    mf4.addNum(-2)
    mf4.addNum(-3)
    assert mf4.findMedian() == -2
    # Test case 5: Single element
    mf5 = MedianFinder()
    mf5.addNum(42)
    assert mf5.findMedian() == 42
    print("All test cases passed!")

if __name__ == "__main__":
    test_median_finder()

File: C++/Heap or Priority Queue/find_median_data_stream.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * MedianFinder:
 * Supports adding numbers from a stream and finding the median efficiently.
 *
 * Approach:
 * - Use two heaps:
 *   - maxHeap (smaller half): a max-heap using priority_queue<int>
 *   - minHeap (larger half): a min-heap using priority_queue<int, vector<int>, greater<int>>
 * - Invariant:
 *   - All elements in maxHeap ≤ all elements in minHeap
 *   - The size difference of the heaps is at most 1
 * - Median:
 *   - If odd total elements: median = top of larger heap
 *   - If even: median = average of tops of both heaps
 *
 * Time Complexity:
 * - addNum: O(log n) (heap push/pop)
 * - findMedian: O(1)
 *
 * Space Complexity:
 * - O(n) to store all numbers
 */

class MedianFinder {
    priority_queue<int> maxHeap; // max-heap (stores smaller half)
    priority_queue<int, vector<int>, greater<int>> minHeap; // min-heap (stores larger half)

public:
    MedianFinder() {
        // Constructor: nothing to initialize explicitly
    }

    void addNum(int num) {
        // Always push to maxHeap first
        maxHeap.push(num);

        // Ensure the maxHeap's largest ≤ minHeap's smallest
        if (!maxHeap.empty() && !minHeap.empty() && maxHeap.top() > minHeap.top()) {
            int val = maxHeap.top(); maxHeap.pop();
            minHeap.push(val);
        }

        // Balance the sizes: maxHeap can have at most 1 more than minHeap
        if (maxHeap.size() > minHeap.size() + 1) {
            int val = maxHeap.top(); maxHeap.pop();
            minHeap.push(val);
        }

        if (minHeap.size() > maxHeap.size() + 1) {
            int val = minHeap.top(); minHeap.pop();
            maxHeap.push(val);
        }
    }

    double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.top();
        }
        if (minHeap.size() > maxHeap.size()) {
            return minHeap.top();
        }

        // If equal, return average of two middle numbers
        return (maxHeap.top() + minHeap.top()) / 2.0;
    }
};

/*
-----------------------------------------------
📄 Line-by-Line C++ Syntax Explanation:
-----------------------------------------------

#include <bits/stdc++.h>
    → Includes all standard C++ headers.

using namespace std;
    → Avoids writing std:: everywhere.

class MedianFinder { ... };
    → Defines the MedianFinder class.

priority_queue<int> maxHeap;
    → A max-heap storing the smaller half of numbers.

priority_queue<int, vector<int>, greater<int>> minHeap;
    → A min-heap storing the larger half of numbers.

MedianFinder() { ... }
    → Constructor of the class, runs when you create an object.

void addNum(int num) { ... }
    → Method to add a number into the data structure.

maxHeap.push(num);
    → Push `num` into the maxHeap.

if (!maxHeap.empty() && !minHeap.empty() && maxHeap.top() > minHeap.top()) { ... }
    → Ensures the two halves are properly ordered.

int val = maxHeap.top(); maxHeap.pop();
    → Extract the largest element from maxHeap.

minHeap.push(val);
    → Push that element into minHeap.

if (maxHeap.size() > minHeap.size() + 1) { ... }
    → Balance the heaps if maxHeap has more than one extra element.

double findMedian() { ... }
    → Method to compute and return the median of all numbers added so far.

if (maxHeap.size() > minHeap.size())
    → If maxHeap has more elements, its top is the median.

if (minHeap.size() > maxHeap.size())
    → If minHeap has more elements, its top is the median.

return (maxHeap.top() + minHeap.top()) / 2.0;
    → If both have equal elements, average of tops of both is the median.

-----------------------------------------------
⏱ Time & Space:
- addNum: O(log n)
- findMedian: O(1)
- Space: O(n) (all numbers stored in heaps)
-----------------------------------------------
*/
